### PR TITLE
Slots inprovment

### PR DESCRIPTION
**Pull request by Ryex**
improve the slots interface:
  - [x] make the fields collapsible
  - [x] display quantity with image
  - [x] let user search for and select occupants as well as change quantity
  - [x] try to make it as compact as possible (slot filter?)
  - [x] ~~pageify if number of slots is large (vending machine)?~~

### CODE REVIEW

**Comment by Ryex** [comment]
@ComputerDruid @Emilgardis 

next update is pretty much ready 
if either of you want / can help out with a review that would be great.

File: .github/workflows/devlop-deploy-netlify.yml
Comment (Emilgardis) [inline .github/workflows/devlop-deploy-netlify.yml:1]: Is the filename intentional? `devlop`

File: ic10emu/src/device.rs [hunk 1]
```diff
@@ -549,10 +546,10 @@ pub struct Device {
     pub name_hash: Option<i32>,
     pub prefab: Option<Prefab>,
     pub slots: Vec<Slot>,
-    pub reagents: HashMap<ReagentMode, HashMap<i32, f64>>,
+    pub reagents: BTreeMap<ReagentMode, BTreeMap<i32, f64>>,
     pub ic: Option<u32>,
     pub connections: Vec<Connection>,
-    fields: HashMap<LogicType, LogicField>,
+    fields: BTreeMap<LogicType, LogicField>,
```
Comment (Emilgardis) [inline ic10emu/src/device.rs:552]: nice, also noticed this

File: ic10emu_wasm/src/types.rs [hunk 2]
```diff
@@ -55,5 +63,51 @@ impl From<&ic10emu::device::Slot> for Slot {
...
+//     prefab_name: Option<String>,
+//     #[serde_as(as = "Vec<(_, _)>")]
+//     fields: BTreeMap<LogicType, DeviceLogicField>,
+//     slots: Vec<DeviceSlot>,
+//     #[serde_as(as = "Vec<(_, _)>")]
+//     reagents: BTreeMap<ReagentMode, Vec<(i32, f64)>>,
+//     connections: Vec<Connection>,
+//     ic: Option<u32>,
+// }
+
```
Comment (Emilgardis) [inline ic10emu_wasm/src/types.rs:112]: is this intentionally left out?

File: ic10emu_wasm/src/types.ts [hunk 3]
```diff
@@ -7,19 +7,19 @@ export interface LogicField {
...
+//   readonly quantity: number;
+//   readonly max_quantity: number;
+//   readonly damage: number;
+//   readonly fields: SlotLogicFields;
+// }
+// export interface Slot {
+//   readonly typ: SlotType;
+//   readonly occupant: SlotOccupant | undefined;
+//   readonly fields: SlotLogicFields;
+// }
```
Comment (Emilgardis) [inline ic10emu_wasm/src/types.ts:22]: ditto, is this supposed to be left in?

**Comment by ComputerDruid** [comment]
I should have some time to take a look today, sure!

**Comment by ComputerDruid** [comment]
I tried the demo and came away with some notes:

- You probably want to write patch notes for this still
- I can't change the ReferenceId of a IC10 chip I added to a second CircuitHousing. Not that I know why it would matter what the id of a slot occupant is, but maybe it would matter if it was using an ID I wanted to use for something else?
- wow, adding a stack of coal ore to the input slot of the sorter, and seeing it rendered with the icon and the little 1/50 looks really good, very impressive.
- I can add a second circuit housing and a chip to it, but it still doesn't appear in the drop-down for setting the active IC. Presumably that's just not implemented yet? Also changing the ID of the main Housing makes it so that it doesn't run anymore, but I think that was true before this PR.
- wow, the device search is fast now! Presumably helped by the "too many results" page.

I tried it out with a random file I had open in a tab and was able to add stuff to the import slots and see it react (turning on the silo), so that's a good sign. [link to my test script](https://deploy-preview-30--ic10emu.netlify.app/#H4sIAAAAAAAAA-1c62_bOBL_VwT1yx6gPYh6WcqXQ87pXgI0j62DLRZNUNA2nWgjiT6JSpMt_L_fkNRbciy5Sa7ZFQzkIc6DHM5vZjiS_E29D9WDb6q_SNSDz9_UJbn3F0Q9QJrqL9UDQ1NjcuMnjMR8HGn61o-ha2L4GjjXwOnB74V6YDuamjC8uAP27czjZ_yMn7_4ByIDDnyckIRHnCVm-PJxDbFGPcVrVVPvcZDCf58_q_Hhgvn3mBFV-6Z-zOLPbE0WIlJFSz8mQEAj9UDXVIbjG8IgYtmbzbUGzJf-4i75QFZsELcF3NcbDQLgyo92zvDIv_EZDn7DwT1M8mdkuLrnWkg3-Ax-SeMIQxDVkGdNPMtyHYtf_g9OZiRKaCwYbMNzTUcXI8cEs0wUcm0YsUxjwgc--P9N_eUlju64MNdFyEV2Y-Q3GqQhcDq6rvORU_zwCUwX59ft7PqZz2II7TM_oDAB15xMJrZpo-oQjYEPBs2JZxuOjWw-eP7gs4zLMTzbMSd6eTnjMBzXMpBridV8pDSEtYIcDzmGZXqOYJAiELJt17EdV1iq0Ih03UO2bjtC5aXYFC7nkoRgeFM3_yknA4vCrFgFsoDNBi1mbSwTijxkufDD8_iosEllZ3THgaU4k2LspLANsmzXgL2Um1MM5mvVLdu2PBM5xWi2QY5rm7pj2qU-OX3Ddfn0uXet_SjLs6VvpZH0uaW60f6fA9ciUTN-9XefBEtw-AVd8n8licLNoxwfzo5_ulJnLE4XLI3J7Ojf_PqV-o-rTJRSeExOLC4oTbLqXuaU-bUWcXWLcmJxTTnpoK74ek6cXWqRZi7RWld2vTWHBn05h1MSMKWDq0RqS0c5NAtxELR0VZmkIqFEsil8tFuRhL7CI4Ly7kNBUI8Mit0YLjxWkQ6rvEP2tBitY1KRkFTemXpJIhHS9hBxue0gNerMRVqkdVC33aTFUItlrc1vkdfR3eFZLY4ijLcWWoxUybNomBOLDeTXlEMW0mR9S-Kat1QTS0tBdbDKVOSPthY-pJQMIv8qRX5VYmQXF4u8CVetq4jR9QGMcUoCf4QUxFdoDP0qCqigeeTR4ipKlOVcmRHG_OimQpmzFjo54zxSYr0KjdLfpXeerzIvP7wnMb4hID7gE-NsdT8upAm01-LErymOmM8eK0Ju2J-cNoZJ4GgpBAqZ75SvQKswqoTcduAs_7qKYI8rpqoYrdCZR4NaaOAAAX18x2qapeISZ_UpPC3zIiZJUhcY5AINXSAZfp9e4KbQBIS2LXO-JlG-jqTbeLpyvlika59U9oAbZJdkKfdp0yVcRga5BgK71BaT7OQwOjiSdF7dWz9cByQksMFzOImuM1tmS-gUOngZ1ZjWYw1V8udZQFXi4Nk3wmyPBTQ4nmcNDaGDl1EP_j1WUWd4nkXUZfZdQwbBsnbqmn0BuZKMy9e3y9cb8mtF15MqapQDtVTrryeVVAl76eAqygycJ9fuiFsEyHrtItXXE22RQLNJzCPyZ0W5zIKZM1TyZPkn7ONc5JbyEs-OsJzVqpIin1T8VEzVdwQrPcu_8s9M3R-Qz2AKV5Eo7mVrTx4_eG8PaWqEQ17fF-XrlEYspgFU_uDZKzz_khEUFcjUjxepz45pmkCWB7okoEyKZI9roLyI6U2MwxDPAzK99fmpmfLth-Qqjt-yp7ji89t5xsYh7CM_xAvyL0ySfSR4WdLpG3kajshZGs75-ayD_FPs815CzmN4m-ygnBcIO5QgQX8hTHKMk9sd5D9PLIu3A1zJN0gJ31Gw7DTASbJ76dcb3vOAY1oUyZ6G3Iopt_4ZYV9pfMeNHIkOh5Zt0RGYXN3wc2An3dZzY77D9CuYWejttY39zYYM15rwvoSwxEcC9V9MllJfHy_Ias9eLiCNp0mPNAsklCfV7RjITrwdzn9GI9Jw-G5bcsVDWVr7rJ6zW26Z_Pd37jvad4OrzbpddpcbNQ0Ijk9JSOM2KDoZ3scx7ecD7x_WgKApTSPWi_4kHEb_gS7u-i_0lPdPelOfRwNooYDvT90fQ73RWjbyhgQ5PYP2Co690YKcLHvSDwoFJbCtAti1xtKI7RHbI7bfOLbtAtv1PvAzgVtUyPbzl6r9q0576K4ZaGK5nme5L1p2oqzuHMPcGOb2DHPoLYc59LphzinCXPUO1ljBjND-IaE9VjD9oT3pqGCqd5G3g7wk-LFAvu13T_D_QJie0nCeJuIxoTcIbPzgh2nYrx4dFgSGAbs3XIsHpXI-cVtlZyWqZwiEoz3th1VOOcXxnEZHPn3wt6y9k-34cRnTmy3r7-SQt7D3VCeZ91XKi4W9-dLkfJ-5nj887qHxggZBynBPIFQYZUsnID365DljT6ugBsdgewy1xB42kCxkKVJHf7YZI7hfYBDke9h4wIQIHJoj1lM2Wfhr0iOW7JfEB9w7kAyVW447FBiTjIMyHJzS3cas35xwG1VC8xGw7VVC4_GySrWw9z2jLKH3TdQDs-iQvDUgt-TPzQ7JLWNiGRPL8yeWJseYWF4osfwwwX9g6Jfk-aP7vUnzpyWHnD695k3v7Pn2rY2lfPxVj5zfw8KT1pGf3D3z4bZXh-owWg481I5tp-9vO6VsnXZP-bUaT5U3WV63mYT0jnvdI6RHSI-QfrOQRu17PyOiR0SPiH6ziDa6HloZMT1iesT0m8V0-fR418ud23Fd-T6A727MvlBDdsiz-_kXGuzZZR17p2PvdOydvr3e6aBWaCVqWh1Rs3xZfXvUrH_zyvbA-UrFQ_9a4OUe4Rjy0OWLZN7i22-G3jR72Rw9uLN_LfY9ki5Se1GyfHfS0lzN1jzN0UwNnBjpGhT1SIOz-kRDJrCv-eS-0Ch4BGpTs4Da4d55i8E9g9a3v-SvLebTUM_wWfNLWv5eNNcb8bLqCqcB-xLleEUbTZWv4Z5M-X__A4t6Z6a5TgAA)

Overall the new stuff feels really nice to use. I'm planning to look at the code later today, too.

**Comment by Ryex** [comment]
> I tried the demo and came away with some notes:
> 
>     * You probably want to write patch notes for this still
> 

absolutly

>     * I can't change the ReferenceId of a IC10 chip I added to a second CircuitHousing. Not that I know why it would matter what the id of a slot occupant is, but maybe it would matter if it was using an ID I wanted to use for something else?
> 

 the slot occupant Id change will need to wait for a bigger rewrite I'm planning

>     * wow, adding a stack of coal ore to the input slot of the sorter, and seeing it rendered with the icon and the little 1/50 looks really good, very impressive.
> 

(^^)


>     * I can add a second circuit housing and a chip to it, but it still doesn't appear in the drop-down for setting the active IC. Presumably that's just not implemented yet? Also changing the ID of the main Housing makes it so that it doesn't run anymore, but I think that was true before this PR.

I think I can fix the housing Id change, in theory I can enable multiple IC's at this point but I'd rather wait till after that bigger rewrite.

>     * wow, the device search is fast now! Presumably helped by the "too many results" page.
> 

Yeah, turns out rending that many dom nodes at once is slow, who knew? (I did, I just didn't consider *how* slow)

> Overall the new stuff feels really nice to use. I'm planning to look at the code later today, too.

File: ic10emu/src/device.rs (hunk 1 above)
Comment (Ryex) [reply ic10emu/src/device.rs:552]: As I noted in the commit comment I decided stable field order was necessary. it means less front-end updates and stable serialising

File: .github/workflows/devlop-deploy-netlify.yml
Comment (Ryex) [reply .github/workflows/devlop-deploy-netlify.yml:1]: nope :P

**Comment by ComputerDruid** [comment]
I did read through the rust code but most of this change wasn't rust (and I'm not familiar enough with js to be able to offer much insight, so I didn't try reading that). The bits I read all look good to me!

**Pull request merged**